### PR TITLE
[SPARK-31848][CORE][TEST] DAGSchedulerSuite: Break down the very huge test file

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current `DAGSchedulerSuite` is a very huge test file, we should split this suite into multiple suites which should focus on one or several major features, but not all the related behaviors.
This PR will extract the following features related tests as new suite from the `DAGSchedulerSuite`:

- Barrier execution

- Indeterminate stage

- Resource profile

### Why are the changes needed?
Break down the very huge test files, each test suite should focus on one or several major features, but not all the related behaviors.

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by beliefer** [comment]
cc @Ngone51

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+  /** The list of results that DAGScheduler has collected. */
+  val results = new HashMap[Int, Any]()
+  var failure: Exception = _
+  val jobListener = new JobListener() {
+    override def taskSucceeded(index: Int, result: Any) = results.put(index, result)
+    override def jobFailed(exception: Exception) = { failure = exception }
+  }
+
+  /** A simple helper class for creating custom JobListeners */
+  class SimpleListener extends JobListener {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:272]: This is only used by `DAGSchedulerSuite`. Shall we keep it at `DAGSchedulerSuite`?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+  }
+
+  // Helper function to validate state when creating tests for task failures
+  private def checkStageId(stageId: Int, attempt: Int, stageAttempt: TaskSet): Unit = {
+    assert(stageAttempt.stageId === stageId)
+    assert(stageAttempt.stageAttemptId == attempt)
+  }
+
+  // Helper functions to extract commonly used code in Fetch Failure test cases
+  protected def setupStageAbortTest(sc: SparkContext): Unit = {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:441]: This is only used by `DAGSchedulerSuite` too. Shall we keep it at `DAGSchedulerSuite`?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+  /**
+   * Common code to get the next stage attempt, confirm it's the one we expect, and complete it
+   * successfully.
+   *
+   * @param stageId - The current stageId
+   * @param attemptIdx - The current attempt count
+   * @param numShufflePartitions - The number of partitions in the next stage
+   * @param hostNames - Host on which each task in the task set is executed
+   */
+  protected def completeShuffleMapStageSuccessfully(
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:469]: This is only used by `DAGSchedulerSuite` too. Shall we keep it at `DAGSchedulerSuite`? Could you check others?

File: core/src/test/scala/org/apache/spark/scheduler/IndeterminateStageSuite.scala [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      null))
+    mapOutputTracker.removeOutputsOnHost("hostB")
+
+    assert(scheduler.failedStages.toSeq.map(_.id) == Seq(1, 2))
+    scheduler.resubmitFailedStages()
+
+    def checkAndCompleteRetryStage(
+                                    taskSetIndex: Int,
+                                    stageId: Int,
+                                    shuffleId: Int): Unit = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/IndeterminateStageSuite.scala:79]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala (hunk 1 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:272]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala (hunk 2 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:441]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala (hunk 3 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:469]: `completeShuffleMapStageSuccessfully` is used by `DAGSchedulerSuite`, `BarrierExecutionSuite` and `IndeterminateStageSuite`.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+      Nil
+    }
+  }
+
+  override def toString: String = "DAGSchedulerSuiteRDD " + id
+}
+
+class DAGSchedulerSuiteDummyException extends Exception
+
+class DAGSchedulerTestHelper extends SparkFunSuite with TempLocalSparkContext with TimeLimits {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerTestHelper.scala:119]: Offline discussion between @Ngone51 and me, decided to use `DAGSchedulerTestBase`

**Comment by beliefer** [comment]
@jiangxb1987

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
cc @jiangxb1987

**Pull request closed without merging**